### DISCUSSION TITLE
Jupyter server proxy wrapped applications

### DISCUSSION THREAD

**Discussion by psychemedia** [mmh352/ou-container-builder#5] [General]
In commit https://github.com/mmh352/ou-container-builder/commit/aa08612127a748ab45e96b9ce19e6e5b5d548459 I note that you moved away from `jhsingle-native-proxy` to using `jupyter-server-proxy` and `jupyterhub`. I'm intrigued as to what led you to that decision and what the consequences are?

I probably haven't been following the evolution of [`jupyter_server`](https://github.com/jupyter-server/jupyter_server) properly. Is it now self-contained enough that you can just use it with `jupyter-server-proxy` as a harness to give the life signals that `jhsingle-native-proxy` used to mock, and the `jupyterhub` package then provides that wrapper that makes it play nice with the JupyterHub server? 

Plus, from the server-proxy, you also get a proxy in front of the application?

(By the by, I notice you didn't upgrade the version number when you made that change, which possibly merits it given you changed the machinery for running the container in JupyterHub...)

**Comment by mmh352** [comment]
I was having trouble with jhsingle-native-proxy handling both scenarios: running behind JH and running directly through docker. I then had another look at the state of jupyter_server and saw that they had cleanly refactored the notebook stuff out of the core server, so I decided to switch to that. The big advantage of it is that it simplifies the code structure, because it uses the same base setup, regardless of whether a notebook or lab server is also desired.

It also means that the application behind the jupyter-server-proxy can always work on the assumption that it is behind the jupyter-server-proxy and gets the appropriate url path prefix passed to it, regardless of whether it is running directly in the container or behind JH.

I haven't updated the version number, because it is still so early in the dev process any nobody but me and soon, hopefully, you is using it :-). When the code stabilises, then the version numbers will be maintained properly.

**Comment by psychemedia** [comment]
I'm just trying to put a couple of demos together to familiarise myself with the config file and create a couple of containers for SSU to try to run with their set-up.

When I ran the upgrade it didn't, for version number issues. Not a major issue — I'll work from a local synched copy of the repo from now on — but I don't want to fall into trap of giving things to SSU to test that are already deprecated.

**Reply by mmh352** [reply]
There are two things that are on my todo list. First, setup a proper rtd docs site and second a proper release flow. As you are now working on this, I'll push them higher up the list.
